sql/pgwire: fix precision loss in binary decimal decoding

dils2k opened the pull request:
Decimal decoding implementation was ignoring trailing zeros after the decimal point that were implicitly added by the dscale value, resulting precision losses. For example, 1.23000 could be decoded as 1.23. This commit fixes it by adding missing zeros.

Release note: None
Epic: CRDB-57092
Fixes: #158093

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158703)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dils2k)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 656 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-OfdyPJ977PLxA_BRA5u:-OfdyPJ977PLxA_BRA5v:b6yox07) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L656)):*
> ```Go
> 				// * if the digits we have in the buffer on the RHS, as calculated above,
> 				// is larger than our calculated dscale, truncate our buffer to match the
> 				// desired dscale.
> ```

Let's make sure to update this comment to describe the case when the buffer is padded with `0`'s and give an example as well.
___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofdya7fD4c4HBrrog0L:-Ofdya7fD4c4HBrrog0M:be8c3rk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L657)):*
> ```Go
> 				// is larger than our calculated dscale, truncate our buffer to match the
> 				// desired dscale.
> 				dscale := (alloc.pgNum.Ndigits - (alloc.pgNum.Weight + 1)) * PGDecDigits
> ```

I think we should go a bit further with this change. This code block is basically just modifying `dscale` to be equal to `alloc.pgNum.Dscale`, and padding or truncating the buffer to compensate for too many or too few trailing zeros in the logic above.

We should get rid of the logic that modifies `dscale` in this block, and just use `alloc.pgNum.Dscale` below when setting the decimal's exponent. `dscale` (which we could rename to `curDScale` or something) will then only be used to determine how much to pad or truncate the buffer.
___
*[`pkg/sql/pgwire/testdata/pgtest/decimal` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofe-011-nuvf-QTk6Ab:-Ofe-011-nuvf-QTk6Ac:bcavw47) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/testdata/pgtest/decimal#L48)):*
> ```
> # CockroachDB intentionally uses exponents for decimals like 1E+4, as
> # oppposed to Postgres, which returns 10000.
> until crdb_only
> ```

I expect that we'll no longer differ from postgres for this case. You can get rid of this test case entirely, and remove the `noncrdb_only` flag from the case below.


<!-- Sent from Reviewable.io -->

dils2k reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158703)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofdya7fD4c4HBrrog0L:-OfhGuS5CQtSKA7ShZhW:be3q1zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L657)):*
> This code block is basically just modifying `dscale` to be equal to `alloc.pgNum.Dscale`

Not always, `dscale` can be negative, resulting a positive exponent and implicit zeros before the decimal point. But I can't see why we can't simply append these zeros to `decDigits` and *always* use `-alloc.pgNum.Dscale` as the exponent, other than for some compression reasons. If it's ok, then I can implement it this way.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> @DrewKimball reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158703)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dils2k)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofdya7fD4c4HBrrog0L:-OfjNdi98pTtkmOiZqjs:b-ya304q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L657)):*
>Not always, `dscale` can be negative, resulting a positive exponent and implicit zeros before the decimal point. But I can't see why we can't simply append these zeros to `decDigits` and _always_ use `-alloc.pgNum.Dscale` as the exponent, other than for some compression reasons. If it's ok, then I can implement it this way.

Good point about negative values, although it turns out we don't actually support such types in SQL yet (https://github.com/cockroachdb/cockroach/issues/158883) and the binary format decoding logic is quite broken when it comes to negative scale: https://github.com/cockroachdb/cockroach/issues/158884.

WRT implicit vs explicit zeros, CRDB represents decimal values differently than Postgres, and doesn't store a separate "precision" value. If we don't append the trailing zeros to the buffer here, the resulting decimal will have a lower precision than was specified by the binary format.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
could we get a test for this in `pkg/sql/pgwire/testdata/encodings.json`, which is used by: https://github.com/cockroachdb/cockroach/blob/33eec654c381339e5bb316939d5ab7f33ae70225/pkg/sql/pgwire/encoding_test.go#L119-L125

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158703)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dils2k)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofdya7fD4c4HBrrog0L:-OfkFdW7E0kjN-909YZ1:b-vooxp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L657)):*
>When I mentioned `dscale` I was referring to the variable

Ah, I missed this. It turns out there's some missing handling for negative `alloc.pgNum.Dscale`, but that can be ignored here.

Can you add a test with this decimal in binary format: `00010002000000030001` (nDigits=1, weight=2, Dscale=3)?
Postgres gives the result: `100000000.000`. My suspicion is that the current implementation will drop the trailing zeros after the decimal point.

I do think we could avoid the padding in the case when  `alloc.pgNum.Dscale` is zero though - in that case, I don't think the trailing zeros are significant.


<!-- Sent from Reviewable.io -->

dils2k commented:
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofdya7fD4c4HBrrog0L:-OfkRFGe3dtAPLr8dLMj:btecul2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L657)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>When I mentioned `dscale` I was referring to the variable

Ah, I missed this. It turns out there's some missing handling for negative `alloc.pgNum.Dscale`, but that can be ignored here.

Can you add a test with this decimal in binary format: `00010002000000030001` (nDigits=1, weight=2, Dscale=3)?
Postgres gives the result: `100000000.000`. My suspicion is that the current implementation will drop the trailing zeros after the decimal point.

I do think we could avoid the padding in the case when  `alloc.pgNum.Dscale` is zero though - in that case, I don't think the trailing zeros are significant.
</blockquote></details>

There is a similar test case:
https://github.com/cockroachdb/cockroach/pull/158703/files#diff-370b733141bbee8d3683f18556f8e15e6e7581118d1e0c48192c9206ada643ecR105-R118

The only difference is that the weight is equal to 1 instead of 2. My current implementation gives: `1.000E+4`.  I suspect that your example will produce `1.000E+8`, I will add your test case too. 


<!-- Sent from Reviewable.io -->

dils2k commented:
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofdya7fD4c4HBrrog0L:-OfkWB7W2324LcqWbmvU:bcjpoas) ([raw file](https://github.com/cockroachdb/cockroach/blob/1598624fbaffce33142349fc3846219511a29e90/pkg/sql/pgwire/pgwirebase/encoding.go#L657)):*
> I will add your test case too. 

added *


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158703)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dils2k)

___
*[`pkg/sql/pgwire/testdata/pgtest/decimal` line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-Ofkf_7F2QyTa3TRW6Ft:-Ofkf_7F2QyTa3TRW6Fu:b-pbzmu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5a62b67966d768aa8c0529f85335e314edcd8f6/pkg/sql/pgwire/testdata/pgtest/decimal#L119)):*
> ```
> {"Type":"ParseComplete"}
> {"Type":"BindComplete"}
> {"Type":"DataRow","Values":[{"text":"1.000E+4"},{"text":"1.000E+8"}]}
> ```

Here's the result Postgres gives:
```
{"Type":"DataRow","Values":[{"text":"10000.000"},{"text":"100000000.000"}]}
```
You can run this test against a local instance directly with an invocation like:
```
./dev test pkg/sql/pgwire -f TestPGTest/decimal --test-args "-addr=localhost:5432 -user=postgres" --rewrite
```


<!-- Sent from Reviewable.io -->

dils2k committed:
sql/pgwire: fix precision loss in binary decimal decoding

Decimal decoding implementation was ignoring trailing
zeros after the decimal point that were implicitly added
by the dscale value, resulting precision losses. For
example, 1.23000 could be decoded as 1.23. This commit
fixes it by adding missing zeros.

Release note: None
Epic: CRDB-57092
Fixes: #158093

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nicely done.

> @DrewKimball reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158703#-:-OgTo90oAK4TBoDo6qGM:b-msncor)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dils2k)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
@DrewKimball thoughts on backporting this to 25.4 and 26.1?
